File: core/LargestPalindromeProduct/LargestPalindromeProduct.java
package com.dbc;

public class LargestPalindromeProduct {
    public int largestPalindrome(int n) {
        if (n == 1) {
            return 9;
        }
        int upper = (int) Math.pow(10, n) - 1;
        int ans = 0;
        for (int left = upper; ans == 0; --left) { // 枚举回文数的左半部分
            long p = left;
            for (int x = left; x > 0; x /= 10) {
                p = p * 10 + x % 10; // 翻转左半部分到其自身末尾，构造回文数 p
            }
            for (long x = upper; x * x >= p; --x) {
                if (p % x == 0) { // x 是 p 的因子
                    ans = (int) (p % 1337);
                    break;
                }
            }
        }
        return ans;
    }
}

File: core/LargestPalindromeProduct/LargestPalindromeProduct.py
def largestPalindrome(n: int) -> int:
    if n == 1:
        return 9
    max_num = 10 ** n - 1
    for left in range(max_num, max_num // 10, -1):
        p, x = left, left
        while x:
            p = p * 10 + x % 10
            x //= 10
        x = max_num
        while x * x >= p:
            if p % x == 0:
                return p % 1377
            x -= 1

    return -1


if __name__ == '__main__':
    print(largestPalindrome(2))
